### ISSUE TITLE
llvm: LLVM_LINK_LLVM_DYLIB=ON makes libclang.dylib not self-contained

### ISSUE THREAD

**Issue by ericvw** [Homebrew/homebrew-core#57941]
- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [x] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
- [x] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

## What you were trying to do (and why)

I was building [YouCompleteMe](https://github.com/ycm-core/YouCompleteMe) for `vim` because I recently upgraded my vim plugins and wanted to build against llvm10.  However, I noticed that llvm10 now how has `libclang.dylib` -> `@rpath/libLLVM.dylib`.  YouCompleteMe's build process copies `libclang.dylib` to a local directory within the plugin, which causes the loader unable to find `libLLVM.dylib` when loading the plugin at `vim` startup.

## What happened (include command output)

After, launching `vim`, the YCM logs indicate the following…

<details>
  <summary>Command output</summary>
  <pre>
OSError: dlopen(/Users/eric/.vim/plugged/YouCompleteMe/third_party/ycmd/third_party/clang/lib/libclang.dylib, 6): Library not loaded: @rpath/libLLVM.dylib
  Referenced from: /Users/eric/.vim/plugged/YouCompleteMe/third_party/ycmd/third_party/clang/lib/libclang.dylib
  Reason: image not found
  </pre>
</details>

## What you expected to happen

No errors when launching `vim` after building the YouCompleteMe plugin.

## Step-by-step reproduction instructions (by running `brew install` commands)

These are the steps I have been using for the past few years:

```
brew install vim llvm cmake python
mkdir ycm_build
cd ycm_build
cmake -S ~/.vim/plugged/YouCompleteMe/third_party/ycmd/cpp -B . -DCMAKE_C_COMPILER=/usr/local/opt/llvm/bin/clang -DCMAKE_CXX_COMPILER=/usr/local/opt/llvm/bin/clang++ -DPATH_TO_LLVM_ROOT=/usr/local/opt/llvm
cmake --build . --target ycm_core
```

My investigation led me to #52352 (cc: @Bo98), which introduced the change in behavior, which I understand gives space savings.  I double-checked on https://releases.llvm.org/download.html to see how the prebuilt macOS libraries were built and `libclang.dylb` is not dynamically linked against `libLLVM.dylib`. 

```
$ otool -L clang+llvm-10.0.0-x86_64-apple-darwin/lib/libclang.dylib
clang+llvm-10.0.0-x86_64-apple-darwin/lib/libclang.dylib:
        @rpath/libclang.dylib (compatibility version 1.0.0, current version 10.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
        /usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
        /usr/lib/libncurses.5.4.dylib (compatibility version 5.4.0, current version 5.4.0)
        /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 902.1.0)

$ otool -L /usr/local/opt/llvm@10/lib/libclang.dylib
/usr/local/opt/llvm@10/lib/libclang.dylib:
        /usr/local/opt/llvm/lib/libclang.dylib (compatibility version 1.0.0, current version 10.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.0.0)
        /usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
        /usr/lib/libncurses.5.4.dylib (compatibility version 5.4.0, current version 5.4.0)
        @rpath/libLLVM.dylib (compatibility version 1.0.0, current version 10.0.0)  
        /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 800.7.0)
```

I erred on the side of caution by creating an issue in case discussion was needed instead of being presumptuous with a pull request.

**Comment by SMillerDev** [comment]
This is not a homebrew issue and given the benefits we're unlikely to roll back the change in homebrew.

**Comment by Bo98** [comment]
You should never assume dylibs are self contained. If you use CMake, consider using the provided LLVM CMake modules to return the correct libraries. `llvm-config` may also work. I'm aware of a couple bugs due to be fixed in 10.0.1

Homebrew is _not_ the only package managger using LLVM_LINK_LLVM_DYLIB - Debian etc have been doing so for years so that software may not work there either.

**Comment by ericvw** [comment]
Thanks for the responses, @SMillerDev and @Bo98.

I'll chase this down on the `vim` YouCompleteMe side of things.  @Bo98, thanks for the pointer to look at the LLVM CMake modules.
